### ISSUE TITLE
Removing intl.locale.requested creates inconsistency in the Firefox interface

### ISSUE THREAD

**Issue by Dupond** [ghacksuserjs/ghacks-user.js#952]
Hello,

This is just a little remark to let you know what can be experimented with a non-english Firefox, in case you wouldn't already know (so this is not a complaint at all).

With the `ghacks-user.js` file, all the webpages, as well as my Firefox interface, are in english language, and this is fine (as I assume otherwise it could have some consequences regarding privacy, finger-printing, etc.).

However, this is not my system language (I use Arch Linux with a non-english locale). As a consequence, a few addons displayed in the "Add-ons Manager" have a non-english name and description ; this is bad at least for one reason: the name of the addon is not the same as in most tutorials. For example, for me, the "Skip Redirect" add-on is displayed as "Passe Redirection" (which I assume is a very bad automatic translation in french language) ; this addon, among others, is recommended for example by the 12bytes.org "Configuration Guide for Privacy and Performance", but I've hard a hard time to find out that my addon was the one referenced to by this guide. Furthermore, here's the worse: the name of the addon is not even the same as the name displayed on the addons.mozilla.org webpage, so that when I enter it in the "Search" field it simply finds nothing!

So I find it much more consistent to display everything in english. This is a personal choice, of course, but it eases my life also (well, a very little, but nonetheless ;)

With just a clean Firefox profile and the `ghacks-user.js` file, I have an empty entry in about:config called `intl.locale.requested`. This preference has been removed from the `ghacks-user.js` file a while ago (precisely in #811). I found out that setting it back to `en-US` allows me to solve the problem, having everything, including the addon names and descriptions, displayed in english.

Maybe a note could simply be added in the user.js file to let the user know about this?

However, I'm concerned about the consequences that setting back this pref could have about finger-printing: would it be trackable? Would it have any other negative downside?

I really hope I'm understandable (as you'll certainly have noticed,english is not my primary language, so it took me a looong time to write this issue ;)

Have a nice day!

Best regards,
Dupond

**Comment by Thorin-Oakenpants** [comment]
> With just a clean Firefox profile and the ghacks-user.js file

So your Firefox build is not en-US (?), but you changed the web content language to en-US (that's cool), but you also changed the "app language" to en-US (which governs everything in the UI: and I guess extensions, about pages etc). Is that right?

If so, that's a pretty edge case IMO: just install and use the en-US bundle.

That said, this is more likely a bug (in how extensions get their app language) but I'd have to dig deeper to know for sure: and it's not something I know a lot about. Long term I believe the goal is to have a single release and the language is added on the fly: e.g. Tor Browser wouldn't have to build 35 versions for each architecture and platform and related incremental updates.

---

What happens in a brand new profile for your non en-US build without the ghacks user.js. That is
- lets say you're using `fr` (french)
- open with a new profile
- app language should be in french: add Passe Redirection
- change app language to en-US: may require a restart
- app should now be in english
- what does the extension show: Passe Redirection or Skip Redirect?

I'll test this later on, if you don't get around to it

**Comment by Dupond** [comment]
No, sorry I should have been more precised: since we want to have everything, included the interface, in english language, for ease of use and privacy concerns, I haven't installed any language pack. I just have installed the Firefox package.

So I've made a few tests: with a fresh profile and no user.js, the Firefox interface defaults (obviously) to english language. I install the Skip Redirect addon from AMO (from this page exactly: https://addons.mozilla.org/en-US/firefox/search/?q=skip%20redirect&platform=Linux&appver=77.0.1 since this is the page that is displayed when searching for "Skip Redirect" in the "Search" bar of the Firefox Addon Manager): the addon is added to my Addon Manager with the french name "Passe Redirection", and its description is in french. Same thing for uMatrix: I add it from AMO, and its description is in french.

That's why I thought that Firefox had access to my system language ; but as soon as I add the `intl.locale.requested" preference set to `en-US`, everything, including the addon description, is in english.

**Comment by Thorin-Oakenpants** [comment]
There's three things going on here
- system locale/language
- package language (maybe)
- app language

I don't have a non en-US OS. But that's OK for now : I don't want to simulate en-US yet, I'm interested in what happens when you change the app language

77 on top, nightly 79 on bottom
![hmm](https://user-images.githubusercontent.com/16656956/84060548-f26db800-a9ab-11ea-91ed-edd023359b05.png)

The option to change the app language is gone, but I'm not sure if this is just my a Nightly thing

PS: another thing you can do, is go to about:support and at the very bottom are seven results for internationalization & localization: the only ones for me that aren't en-US are the regional preferences (which match the option in the pictures above for how to format dates, times etc)

**Comment by Dupond** [comment]
Not sur to understand what you want : how would I change the app language, since I don't have any language pack installed ? Or do you want me to install one and see what happens ?

I'm using Firefox 77.0.1, so my "language" section is exactly the same as yours: clicking on "Set Alternatives..." only shows one line (which is "English (United States)"), and the "Choose button" for the "Choose your preferred language for displaying pages" option shows two lines: "English (United States) [en-us]" and "English [en]".

Here is what my `about:support` page displays in the "Internationalization & Localization" section:

> Application Settings
> Requested Locales 	["en-US"]
> Available Locales 	["en-US"]
> App Locales 	["en-US"]
> Regional Preferences 	["en-US"]
> Default Locale 	"en-US"
> Operating System
> System Locales 	["fr-FR"]
> Regional Preferences 	["fr-FR"]

I hope this helps a little bit...

**Comment by Thorin-Oakenpants** [comment]
> Not sure to understand what you want : how would I change the app language, since I don't have any language pack installed ? Or do you want me to install one and see what happens ?

No. You're fine. I can test things from my end at this stage. You normally change the app language in that pic I posted in 77 where it says "choose the language used to display menus..." - but you don't need to

> I hope this helps a little bit...

Probably :) I'm not an expert on how Firefox uses all this stuff. To me it seems like a bug: because in both you have `app locale` as en-US and I would have thought extensions would use that, but clearly they use `requested locale` (as far as I can tell) - maybe that's by design.

**Comment by Thorin-Oakenpants** [comment]
Sorry, just to clarify (but you don't need to do it)

Changing the app language is right under the sentence that says "choose the languages used to display menus..." where there is a combobox, which if you click it shows you a list of all your available languages

I only have en-US
![lang1](https://user-images.githubusercontent.com/16656956/84076991-3bcb0100-a9c6-11ea-91b7-53b6389edea3.png)

I click on Search for more languages
![lang2](https://user-images.githubusercontent.com/16656956/84077331-b267fe80-a9c6-11ea-9b5a-583390518c7f.png)

I click on select a language to add... and choose French
![lang3](https://user-images.githubusercontent.com/16656956/84077445-dc212580-a9c6-11ea-91d2-b4f011f76069.png)

and click `Add` - this then **downloads** the app language for French from Mozilla: it actually says "downloading" in the dialog while it goes and fetches it. And you click OK.

I also added Arabic as well. Here's what it looks like now
![lang4](https://user-images.githubusercontent.com/16656956/84077977-a466ad80-a9c7-11ea-8a59-e07b22bbce03.png)

And when I select a different app language, it asks for a restart to apply it
![lang5](https://user-images.githubusercontent.com/16656956/84078074-d0822e80-a9c7-11ea-83d4-70093defa216.png)

---

So basically, Firefox doesn't want to bundle all the translations etc (I think there's 220 odd of them) for every download of Firefox, so instead, you download them as you want them. That's what I mean by package vs app language

**Comment by Thorin-Oakenpants** [comment]
ooh, cool.. and this them under about:addons as entries under `languages` and `dictionaries` where you can remove them
![cool](https://user-images.githubusercontent.com/16656956/84081305-3c1aca80-a9cd-11ea-8068-e019be043f69.png)

**Comment by Thorin-Oakenpants** [comment]
@kkapsner Just checking, but I assume you use [i18n](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/i18n) to determine the locale of CB, right? e.g.
```js
browser.i18n.getUILanguage()
```

**Comment by kkapsner** [comment]
Yes I use this API - but only `browser.i18n.getMessage()`.

The name and the description can be localised by using `__MSG_something_something__` in the manifest.json (https://github.com/kkapsner/CanvasBlocker/blob/master/manifest.json#L4 - CB does not translate the name but the description) - so everything is controlled by Firefox.

**Comment by Thorin-Oakenpants** [comment]
The user.js is immaterial: that's just for changing your web-content language preferences etc, and does not affect app (menu) or the extension language API (li0n)

My system is not en-US, but it is English, but that's also immaterial (as far as I can tell) because at no stage did I ever see English in my test below

---

Control
- FF77.01 French, vanilla profile (no user.js, OS is English)
- add Skip Redirect
- add CanvasBlocker
- everything is in French
![french](https://user-images.githubusercontent.com/16656956/84261593-8904bb00-ab0b-11ea-8f35-f7b65a9c67ab.png)

Test
- add/download Spanish as an app language option
- apply spanish and restart
- expected: everything to be in Spanish
- actual: CanvasBlocker is in Spanish, Skip Redirect is not
![spanish](https://user-images.githubusercontent.com/16656956/84261698-c10bfe00-ab0b-11ea-855a-55098bd0e731.png)

---

Conclusion
- the web extension APIs are being used successfully in CanvasBlocker, but not in Skip Redirect (or the other addons you mentioned). Whatever they're doing that is different is the issue

**If** Skip Redirect is using `li0n` then I would consider this a bug / inconsistent application (and I can contact the right Moz dev). Clearly Skip Redirect has several languages, it's not like it's not catering to locale at all.

@Dupond  You mentioned several extensions: can you name them please

pinging @sblask

**Comment by Thorin-Oakenpants** [comment]
great: https://github.com/sblask/webextension-skip-redirect/tree/master/_locales

I'll retest by adding and flipping to german

**Comment by kkapsner** [comment]
> I didn't choose translating the name.

You choose that it's possible to translate it: https://github.com/sblask/webextension-skip-redirect/blob/master/manifest.json#L35 vs. https://github.com/kkapsner/CanvasBlocker/blob/master/manifest.json#L3
Maybe you should change that...

**Comment by sblask** [comment]
Maybe I should have, but I think it's too late for that now. Apparently it made sense for the translator to translate it and I so far have not had any complaints. But changing the name is disruptive.

**Comment by Dupond** [comment]
Oh. My. God.

I just discovered something: I was looking for the origin of this empty `intl.locale.requested` pref in my browser, since this pref shouldn't exits at all with a default Firefox installation.

It turns out that (as shown in the list above) Arch Linux Firefox package contains a `vendor.js` file (located in `/usr/lib/firefox/browser/defaults/preferences/) with the following content:

```
// Use LANG environment variable to choose locale
pref("intl.locale.requested", "");

// Use system-provided dictionaries
pref("spellchecker.dictionary_path", "/usr/share/hunspell");

// Disable default browser checking.
pref("browser.shell.checkDefaultBrowser", false);

// Don't disable extensions in the application directory
pref("extensions.autoDisableScopes", 11);
```

So the first 2 lines suggested that even an empty pref would be enough to switch locale. And indeed, it is ! As I checked in my `about:support` page, commenting this line is enough to keep `Requested Locales` set to default, which is `["en-US"]`.

So it turns out that this is an Arch-Linux-only problem; but this issue isn't "Removing intl.locale.requested creates inconsistency" anymore; it should rather be "Adding intl.locale.requested (even if empty) creates inconsistency in the Firefox interface"...

**Comment by Thorin-Oakenpants** [comment]
Nah, it's all good - I learned a few things :)

I wonder how many other Linux distro repos do the same thing (and I hope you contact the arch people). That said, maybe Mozilla should fall back to something when it's an `empty string`

Can I close this?

**Comment by Dupond** [comment]
Of course, this should be closed.

And forget.

Forever.

**Comment by Thorin-Oakenpants** [comment]
OMG ... [1537704#c20](https://bugzilla.mozilla.org/show_bug.cgi?id=1537704#c20) ... `und` as a locale ... and I've read this bug before
